src/connectors/impls/grpc/client.rs

```diff
@@ -0,0 +1,208 @@
+
+#[derive(Debug, Clone)]
+pub struct GrpcClient {
+    tx: Sender<EventPayload>,
+    rx: Receiver<EventPayload>,
+    config: Config,
+    event_origin_uri: EventOriginUri,
+    sink: Option<GrpcClientSink>,
+}
+
+#[derive(Debug, Clone)]
+struct GrpcClientSink {
+    grpc_client_handler: Arc<Mutex<Option<tremor_grpc::GrpcClientHandler>>>,
+    error_rx: Receiver<Error>,
+    error_tx: Sender<Error>,
+}
+
+#[derive(Debug, Clone)]
+struct GrpcClientSource {
+    rx: Receiver<EventPayload>,
+    origin_uri: EventOriginUri,
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "grpc_client".into()
+    }
+
+    async fn from_config(
+        &self,
+        _id: &str,
+        config: &Option<OpConfig>,
+    ) -> Result<Box<dyn Connector>> {
+        if let Some(raw_config) = config {
+            let config = Config::new(raw_config)?;
+            let (tx, rx) = unbounded();
+            let event_origin_uri = EventOriginUri {
+                scheme: "tremor-grpc-unary".to_string(),
+                host: config.host.clone(),
+                port: Some(config.port.clone()),
+                path: vec![],
+            };
+            Ok(Box::new(GrpcClient {
+                config,
+                tx,
+                rx,
+                event_origin_uri,
+                sink: None,
+            }))
+        } else {
+            Err(ErrorKind::MissingConfiguration(String::from("GrpcClient")).into())
+        }
+    }
+}
+
+#[async_trait::async_trait()]
+impl Connector for GrpcClient {
+    fn is_structured(&self) -> bool {
+        true
+    }
+
+    async fn create_source(
+        &mut self,
+        ctx: SourceContext,
+        builder: SourceManagerBuilder,
+    ) -> Result<Option<SourceAddr>> {
+        let source = GrpcClientSource {
+            rx: self.rx.clone(),
+            origin_uri: self.event_origin_uri.clone(),
+        };
+        let addr = builder.spawn(source, ctx)?;
+        Ok(Some(addr))
+    }
+
+    async fn create_sink(
+        &mut self,
+        ctx: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        let (error_tx, error_rx) = unbounded();
+        let sink = GrpcClientSink {
+            grpc_client_handler: Arc::new(Mutex::new(None)),
+            error_rx,
+            error_tx,
+        };
+        self.sink = Some(sink.clone());
+        let addr = builder.spawn(sink, ctx)?;
+        Ok(Some(addr))
+    }
+
+    async fn connect(&mut self, _ctx: &ConnectorContext, _attempt: &Attempt) -> Result<bool> {
+        let grpc_client_handler = tremor_grpc::GrpcClientHandler::connect(
+            format!("{}:{}", self.config.host, self.config.port),
+            self.tx.clone(),
+        )
+        .await?;
+        if let Some(sink) = &mut self.sink {
+            let mut client_handler = sink.grpc_client_handler.lock().await;
```

Comment:
sink and source now also have a `connect` method, so you can (and should) move this there, as the clone above will result in operating on two distinct entities